### PR TITLE
Adding date and time input when creating a contest

### PR DESCRIPTION
**Pull request by emmyachieng**

### CODE REVIEW

**Comment by Japhetb** [comment]
https://user-images.githubusercontent.com/52126448/132495285-bf5ed0c1-a51c-483d-9ef6-b37ed231b8a3.mov

**Comment by sigu** [comment]
1. Title still contains the word draft
2. When I submit an empty form (not filling in any detail) I get the following error
```
error] GenServer #PID<0.736.0> terminating
** (MatchError) no match of right hand side value: {:error, :invalid_format}
    (quadquizaminos 0.8.6) lib/quadquizaminos_web/live/contests_live.ex:60: QuadquizaminosWeb.ContestsLive.handle_event/3
    (phoenix_live_view 0.15.5) lib/phoenix_live_view/channel.ex:342: anonymous fn/3 in Phoenix.LiveView.Channel.view_handle_event/3
    (telemetry 0.4.3) /Users/sigu/code/clients/sparrell/quizquadaminos/deps/telemetry/src/telemetry.erl:272: :telemetry.span/3
    (phoenix_live_view 0.15.5) lib/phoenix_live_view/channel.ex:204: Phoenix.LiveView.Channel.handle_info/2
    (stdlib 3.15.1) gen_server.erl:695: :gen_server.try_dispatch/4
    (stdlib 3.15.1) gen_server.erl:771: :gen_server.handle_msg/6
    (stdlib 3.15.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: %Phoenix.Socket.Message{event: "event", join_ref: "4", payload: %{"event" => "save", "type" => "form", "value" => "_csrf_token=GA4BNQxaDh9QIXQiNz1CWiYiLTwKBT1WzEcS5nbQ1S-snnv5iMWKsbv4&name=&contest_date="}, ref: "5", topic: "lv:phx-FqRYaRY4-Mhe6gED"}
```

File: lib/quadquizaminos_web/live/contests_live.ex [hunk 1]
```diff
@@ -1,5 +1,6 @@
 defmodule QuadquizaminosWeb.ContestsLive do
   use Phoenix.LiveView
+  use Phoenix.HTML
```
Comment (sigu) [inline lib/quadquizaminos_web/live/contests_live.ex:3]: You can use QuadquizaminosWeb, :live_view which by default includes these view helpers

File: lib/quadquizaminos_web/live/contests_live.ex [hunk 2]
```diff
@@ -55,10 +56,12 @@ defmodule QuadquizaminosWeb.ContestsLive do
     {:noreply, assign(socket, contests: contests, editing_date?: true)}
   end
 
-  def handle_event("save", %{"key" => "Enter", "value" => contest_name}, socket) do
-    {:noreply, socket |> _create_contest(contest_name)}
+  def handle_event("save", %{"name" => name, "contest_date" => contest_date}, socket) do
```
Comment (sigu) [inline lib/quadquizaminos_web/live/contests_live.ex:59]: What if no date is selected? Should we default to today? Should we just make it null?

File: lib/quadquizaminos_web/live/contests_live.html.leex [hunk 3]
```diff
@@ -4,8 +4,13 @@
         <% else %>
           <%= if @countdown_interval <= 0 || @current_user.admin? do %>
         <h1>Contests</h1>
-
-        <%= raw display_text_input(@current_user.admin?) %>
+        <%= if @current_user.admin? do  %>
+          <%= form_for :contest, "#", [phx_submit: :save] %>
```
Comment (sigu) [inline lib/quadquizaminos_web/live/contests_live.html.leex:8]: Use the contest changeset to validate this form

**Review by sigu — CHANGES_REQUESTED** [review]

**Comment by Japhetb** [comment]
> 1. Title still contains the word draft
> 2. When I submit an empty form (not filling in any detail) I get the following error
> 
> ```
> error] GenServer #PID<0.736.0> terminating
> ** (MatchError) no match of right hand side value: {:error, :invalid_format}
>     (quadquizaminos 0.8.6) lib/quadquizaminos_web/live/contests_live.ex:60: QuadquizaminosWeb.ContestsLive.handle_event/3
>     (phoenix_live_view 0.15.5) lib/phoenix_live_view/channel.ex:342: anonymous fn/3 in Phoenix.LiveView.Channel.view_handle_event/3
>     (telemetry 0.4.3) /Users/sigu/code/clients/sparrell/quizquadaminos/deps/telemetry/src/telemetry.erl:272: :telemetry.span/3
>     (phoenix_live_view 0.15.5) lib/phoenix_live_view/channel.ex:204: Phoenix.LiveView.Channel.handle_info/2
>     (stdlib 3.15.1) gen_server.erl:695: :gen_server.try_dispatch/4
>     (stdlib 3.15.1) gen_server.erl:771: :gen_server.handle_msg/6
>     (stdlib 3.15.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
> Last message: %Phoenix.Socket.Message{event: "event", join_ref: "4", payload: %{"event" => "save", "type" => "form", "value" => "_csrf_token=GA4BNQxaDh9QIXQiNz1CWiYiLTwKBT1WzEcS5nbQ1S-snnv5iMWKsbv4&name=&contest_date="}, ref: "5", topic: "lv:phx-FqRYaRY4-Mhe6gED"}
> ```

worked on

File: lib/quadquizaminos_web/live/contests_live.ex (hunk 1 above)
Comment (Japhetb) [reply lib/quadquizaminos_web/live/contests_live.ex]: Done

File: lib/quadquizaminos_web/live/contests_live.ex (hunk 2 above)
Comment (Japhetb) [reply lib/quadquizaminos_web/live/contests_live.ex:66]: We have used null

File: lib/quadquizaminos_web/live/contests_live.ex [hunk 4]
```diff
@@ -159,10 +157,20 @@ defmodule QuadquizaminosWeb.ContestsLive do
     {:noreply, socket |> assign(current_user: current_user, current_uri: uri)}
   end
 
-  defp _create_contest(socket, contest_name) do
+  defp _create_contest(socket, contest_name, "") do
```
Comment (TraceyOnim) [inline lib/quadquizaminos_web/live/contests_live.ex:160]: You can pass the params received as the second argument , then pattern match the date
```suggestion
  defp _create_contest(socket, %{"contest_date" => ""}= params) do
  params = %{params | "contest_date" => nil}
  case Contests.create_contest(params) do
      {:ok, contest} -> assign(socket, contests: contests ++ [contest])
      _ -> socket
    end
```

File: lib/quadquizaminos_web/live/contests_live.ex [hunk 5]
```diff
@@ -159,10 +157,20 @@ defmodule QuadquizaminosWeb.ContestsLive do
...
     contests = socket.assigns.contests
 
-    case Contests.create_contest(%{name: contest_name}) do
+    case Contests.create_contest(%{name: contest_name, contest_date: nil}) do
+      {:ok, contest} -> assign(socket, contests: contests ++ [contest])
+      _ -> socket
+    end
+  end
+
+  defp _create_contest(socket, contest_name, contest_date) do
```
Comment (TraceyOnim) [inline lib/quadquizaminos_web/live/contests_live.ex:169]:
```suggestion
  defp _create_contest(socket, %{"contest_date" => contest_date}=params) do
  contests = socket.assigns.contests
    {:ok, contest_date, 0} = DateTime.from_iso8601(contest_date <> ":00Z")
    params = %{params | "contest_date" => contest_date}

    case Contests.create_contest(params) do
      {:ok, contest} -> assign(socket, contests: contests ++ [contest])
      _ -> socket
    end

```

**Review by sigu — APPROVED** [review]

**Pull request merged**